contrib/neon/pagestore_smgr.c

```diff
@@ -1277,12 +1293,30 @@ zenith_nblocks(SMgrRelation reln, ForkNumber forknum)
 
 	if (get_cached_relsize(reln->smgr_rnode.node, forknum, &n_blocks))
 	{
-		elog(SmgrTrace, "cached nblocks for %u/%u/%u.%u: %u blocks",
-			 reln->smgr_rnode.node.spcNode,
-			 reln->smgr_rnode.node.dbNode,
-			 reln->smgr_rnode.node.relNode,
-			 forknum, n_blocks);
-		return n_blocks;
+		if (n_blocks == InvalidBlockNumber)
+		{
+			/*
+			 * Now this is surprising: our cache says that the fork does not exist.
+			 * PostgreSQL should not call smgrnblocks() on a non-existent fork,
+			 * that's an error in md.c too. It probably means that there's a bug in
+			 * our relation size caching, and we've forgotten to update it. Warn,
+			 * and fall through to get the real size from the page server.
+			 */
+			elog(WARNING, "smgrnblocks() called on %u/%u/%u.%u, but relsize cache says the fork does not exist",
```

Comment:
Shouldn't we throw an error here to catch a bug sooner?

Reply:
Maybe, although it seems unpleasant to fail the query, just for our debugging purposes. Maybe LOG?